File: PPC/Assignments/KnapsackProblem2/KnapsackProblemKotlin/src/main/kotlin/knapsack/actor/actors/BestOfActor.kt
package knapsack.actor.actors

import Individual
import akka.actor.AbstractActor

class BestOfActor : AbstractActor() {

    data class Request(
        val population: Array<Individual>
    )

    data class Response(val best: Individual)

    override fun createReceive(): Receive {
        return receiveBuilder()
            .match(Request::class.java) { request ->
                val population = request.population

                val best = population.maxByOrNull { it.fitness } ?: population[0]

                sender.tell(Response(best), self)
            }
            .build()
    }
}

File: PPC/Assignments/KnapsackProblem2/KnapsackProblemGo/knapsack/actor/actors/bestOfActor.go
package actors

import (
	"KnapsackProblemGo/knapsack"
	"github.com/asynkron/protoactor-go/actor"
)

type BestOfActor struct{}

type BestOfRequest struct {
	Individuals []*knapsack.Individual
}

type BestOfResponse struct {
	BestIndividual *knapsack.Individual
}

func NewBestOfActor() actor.Actor {
	return &BestOfActor{}
}

func (a *BestOfActor) Receive(context actor.Context) {
	switch msg := context.Message().(type) {
	case *BestOfRequest:
		a.handleRequest(msg, context)
	}
}

func (a *BestOfActor) handleRequest(request *BestOfRequest, context actor.Context) {
	individuals := request.Individuals
	var best *knapsack.Individual
	for _, ind := range individuals {
		if best == nil || ind.Fitness > best.Fitness {
			best = ind
		}
	}
	context.Respond(&BestOfResponse{BestIndividual: best})
}
